### ISSUE TITLE
TX data format

### ISSUE THREAD

**Issue by JohnnyPicnic** [mulcmu/esphome-ge-laundry-uart#6]
Hi, have you been able to send message to you GE devices?
I'm using a USB FTDI Serial adapter which is working great but I haven't figured out how to do hardware control for TX.
While I figure that out I'm trying to figure out the proper format for sending messages.

I believe bf is the knobs on my oven
Use “Upper Oven Remote Enable” ERD  (0x510A) – e.g. write ERD and enable 
	•	Sample data [Source Address, 0xF1 (ERD write), 0x01 (1 ERD), 0x51, 0x0A (ERD), 0x01 (size), 0x01 (enable)]
Does that mean the hex code should be E2 bf F1 01 51 0A 01 01 CRC CRC E3 ?

**Comment by mulcmu** [comment]
Yes, I have been able to send and receive to washer and dryer.  Washer had separate RX & TX lines, full duplex.  There wasn't any traffic on these lines until serial adapter or ESP32 started to request data.  Dryer has RX & TX on same line, half duplex.  The dryer main board seemed to be sending packets to internal wifi module that wasn't installed on my unit.  So I was able to see some traffic without initiating a request.

Your sample [packet ](https://github.com/doitaljosh/ge-appliances-re/blob/master/geabus-packet-structure.png)is missing the size/length and source address.  

E2 Start Byte
BF Destination
0D Size (13 bytes in packet)
BB Source (Address the serial adapter is going to assume)
F1 ERD write
01 Single Erd
51 Erd High byte
0A Erd Low Byte
01 Size of data
01 Data to set
CRC
CRC
E3

0XBF looks like it is the universal wifi module from this [link](https://github.com/doitaljosh/ge-appliances-re/blob/master/reverse-engineering.txt)

Based on my testing the CRC needed to be correct otherwise the appliance would not respond.

Here was packet that should request all of the devices in appliance respond with their firmware revision.  This can be used to figure out which boards in oven are at what address.

`std::vector<uint8_t> fw_broadcast= {0XE2, 0XFF, 0X08, 0XBB, 0X01, 0X9A, 0X85, 0xE3};    //FW broadcast message`

What baud rate was the oven using and did you need to setup the FTDI to invert the serial polarity?

**Comment by JohnnyPicnic** [comment]
Yup the baud rate was standard 19200 and yes I had to invert the signal using FT Prog. I don't know why I'm having such a hard time figuring out half duplex with the FTDI. For your setup it's all done through diodes? 
I'm trying to do everything with Node Red. I've a got CRC 16 function node working and now just need to figure out how to transmit.

**Comment by mulcmu** [comment]
The ESP32 hardware has an option for inverting the serial polarity, this is getting implemented in my setup with the ESPHome configuration.  

The U1 and U2 74LVC2G07 is a non inverting buffer used to shift voltage level from 5V to 3.3V for ESP.  The resistor/diode (R5/D4) was added based on this [article](https://circuitcellar.com/resources/ee-tips/one-wire-rs-232-half-duplex-ee-tip-135/).  The inverted TTL idle state is 0V.  The diode allows the appliance to put 5V on the bus line without the U2A buffer fighting to pull it back down to 0V.

**Comment by JohnnyPicnic** [comment]
Thanks for the link to the article. If I already have the signal inverted and don't need voltage shifting do you think the R5/D4 combo would be all I need? I have no problem with just RX connected but once I add TX the data is echoed and probably corrupted. I didn't' notice it but is there any timing involved in your ESP32 code?
I may move to an ESP32 but would still like to parse all the data through Node Red.

**Comment by mulcmu** [comment]
Nothing too critical timing wise.  The send was done after reading and processing the RX input buffer.  Before sending the packet I just checked to see if any new bytes showed up on RX port and would abort if so.  The dryer would repeat the response 3 times unless the `e1` acknowledge byte was sent promptly.

If your oven has built in wifi, it might be harder to find an open window to TX if there is a lot of other traffic.  I don't recall seeing where any timing was discussed in the other related projects.

**Comment by JohnnyPicnic** [comment]
I made the circuit you used for your board and I think I'm getting a response to the command you suggested.
TX: 
0XE2, 0XFF, 0X08, 0XBB, 0X01, 0X9A, 0X85, 0xE3
RX:
e2bb0cbf 01 00 01 05 6c 25 ab e3
e2bb0cbc 01 00 01 00 14 eb 21 e3
e2bb0cbf 01 00 01 05 6c 25 ab e3
e2bb0cbf 01 00 01 05 6c 25 ab e3
e2bb0cbc 01 00 01 00 14 eb 21 e3
e2bb0cbc 01 00 01 00 14 eb 21 e3
e2bb0cbc 01 00 01 00 14 eb 21 e3

**Comment by mulcmu** [comment]
Have you made any more progress getting FTDI to work with oven?

**Comment by JohnnyPicnic** [comment]
Not really. No other command seemed to give me a response. 
I ordered a FT260 HID-class USB to I2C/UART module in hopes of emulating the Green Bean.

**Comment by mulcmu** [comment]
I have a few extra of the Rev 0.3 pcb built if you want one of those to try.  You'll just need to supply the ESP32.

**Comment by JohnnyPicnic** [comment]
Thats a really generous offer. How much? Shipping would be to Edmonton, Alberta, Canada.

**Comment by mulcmu** [comment]
No charge for PCB just cover the shipping.  Email is mulcmu at yahoo.

**Comment by mguaylam** [comment]
Have you guys seen this? https://github.com/geappliances/tiny-gea3-api

**Comment by mulcmu** [comment]
Thanks.  They also posted the erd list in another repository: https://github.com/geappliances/public-appliance-api-documentation

**Comment by mguaylam** [comment]
@JohnnyPicnic should we close this issue since it's now well documented in other repo?

**Comment by JohnnyPicnic** [comment]
Yup, been a while since I've looked at this. Your post reminded me to see if any of the new adapters are available so I just ordered one. Are you using your board or the one from FirstBuild?

**Comment by mguaylam** [comment]
There's two boards I'm aware of. The one of @mulcmu and GE. I still need to try the mulcmu board with a fix but other users are claiming it works.
For the projects, there's a couple but the only ones on ESPHome which I think are the best use for it are @mulcmu project (this) and mine which is feature complete. 

I was telling you because I'm cleaning the issues I'm involved in.
